Fix - VR Month view handling of sticky events

lucatume opened the pull request:
Tickets:
* https://central.tri.be/issues/133303 
* https://central.tri.be/issues/136021
* https://central.tri.be/issues/134827
* https://central.tri.be/issues/133304
* https://central.tri.be/issues/129762

Based on: https://github.com/moderntribe/tribe-common/pull/1184

Screencap: [link](https://drive.google.com/open?id=10dqzaGfuECbzPfk1V_dWYLtjfyRZExul&authuser=luca@tri.be&usp=drive_fs)

This is the "fix" to how Month View handles Sticky events.

I'm not sure, though, if the exp. behavior is the correct one or not.
In v2 of the Month View we've got two event "areas":

1. the stack (all-day and multi-day events, by default)
2. the cell (everything else)

The solution I've implemented will keep the above distinction, but will order events by `menu_order` first (what we set to make events "sticky").
So:

* an sticky (all-day/multi-day) event will be the first in the stack, among the available positions
* a sticky normal event will be the first, in the cell, among the cell events

With two diff. event areas and the concept of the stack, I'm not sure this is how we want to handle this.
Though the code here will provide a base for whatever we need later.

tr1b0t commented on src/Tribe/Repositories/Event.php:
```diff
@@ -1465,26 +1463,48 @@ public function handle_order_by( $order_by ) {
-			$this->filter_query->orderby( $timestamp_key );
-		} else {
-			$this->query_args['orderby'] = $order_by;
+		$after    = false;
```
:warning: Equals sign not aligned with surrounding assignments; expected 1 space but found 4 spaces

<sub>hash: 86f61c31abdb196da4f71c2c7c155000</sub>

tr1b0t commented on src/Tribe/Repositories/Event.php:
```diff
@@ -1465,26 +1463,48 @@ public function handle_order_by( $order_by ) {
-		} else {
-			$this->query_args['orderby'] = $order_by;
+		$after    = false;
+		$loop = 0;
```
:warning: Equals sign not aligned with surrounding assignments; expected 2 spaces but found 1 space

<sub>hash: f7c65349af9676f39e377d2ec69c62e8</sub>

tr1b0t commented on src/Tribe/Repositories/Event.php:
```diff
@@ -1465,26 +1463,48 @@ public function handle_order_by( $order_by ) {
+					$this->filter_query->orderby( $timestamp_key, null, null, $after );
+					break;
+				default:
+					$after = $after || $loop === 1;
```
:no_entry_sign: Use Yoda Condition checks, you must.

<sub>hash: 9006b2afabb429c7125e3300dbbe0c84</sub>

tr1b0t commented on src/Tribe/Repositories/Event.php:
```diff
@@ -1465,26 +1463,48 @@ public function handle_order_by( $order_by ) {
+					} else {
+						// Make sure all `orderby` clauses have the shape `<orderby> => <order>`.
+						$this->query_args['orderby'] = (array) $this->query_args['orderby'];
+						array_reduce( $this->query_args['orderby'], static function ( array $q, $v, $k ) {
```
* :no_entry_sign: Opening parenthesis of a multi-line function call must be the last content on the line
* :no_entry_sign: Only one argument is allowed per line in a multi-line function call

<sub>hash: 98d40be0d23e09e215fb69056e09306f</sub>

tr1b0t commented on src/Tribe/Repositories/Event.php:
```diff
@@ -1465,26 +1463,48 @@ public function handle_order_by( $order_by ) {
+						array_reduce( $this->query_args['orderby'], static function ( array $q, $v, $k ) {
+							$order_by        = is_numeric( $k ) ? $v : $k;
+							$order           = is_numeric( $k ) ? 'ASC' : $v;
+							$q [ $order_by ] = $order;
```
:no_entry_sign: Space found before square bracket; expected "$q[" but found "$q ["

<sub>hash: d5bdb95c01c96048ee80114b94b83e7a</sub>

tr1b0t commented on src/Tribe/Repositories/Event.php:
```diff
@@ -1465,26 +1463,48 @@ public function handle_order_by( $order_by ) {
+							$q [ $order_by ] = $order;
+
+							return $q;
+						}, [] );
```
* :no_entry_sign: Only one argument is allowed per line in a multi-line function call
* :no_entry_sign: Closing parenthesis of a multi-line function call must be on a line by itself

<sub>hash: 35a9646658a4c71d71dc01089d6cb05d</sub>

bordoni commented:
Sorry tests are not passing.

tr1b0t commented on src/Tribe/Repositories/Event.php:
```diff
@@ -1463,26 +1460,49 @@ public function handle_order_by( $order_by ) {
-			$this->filter_query->orderby( $timestamp_key );
-		} else {
-			$this->query_args['orderby'] = $order_by;
+		$after    = false;
```
:warning: Equals sign not aligned with surrounding assignments; expected 1 space but found 4 spaces

<sub>hash: 6e57fc0e7122e835cf859502dadaba6d</sub>

tr1b0t commented on src/Tribe/Repositories/Event.php:
```diff
@@ -1463,26 +1460,49 @@ public function handle_order_by( $order_by ) {
-		} else {
-			$this->query_args['orderby'] = $order_by;
+		$after    = false;
+		$loop = 0;
```
:warning: Equals sign not aligned with surrounding assignments; expected 2 spaces but found 1 space

<sub>hash: 32850f02c541ba0415765bce00557e47</sub>

tr1b0t commented on src/Tribe/Repositories/Event.php:
```diff
@@ -1463,26 +1460,49 @@ public function handle_order_by( $order_by ) {
+					$this->filter_query->orderby( [ $timestamp_key => $default_order ], null, null, $after );
+					break;
+				default:
+					$after = $after || $loop === 1;
```
:no_entry_sign: Use Yoda Condition checks, you must.

<sub>hash: 73015339f395aa39a4ce7ee1e33c2010</sub>

tr1b0t commented on src/Tribe/Repositories/Event.php:
```diff
@@ -1463,26 +1460,49 @@ public function handle_order_by( $order_by ) {
+					if ( empty( $this->query_args['orderby'] ) ) {
+						$this->query_args['orderby'] = [ $order_by => $order ];
+					} else {
+						$add           = [ $order_by => $order ];
```
:warning: Equals sign not aligned correctly; expected 1 space but found 11 spaces

<sub>hash: 249e4376bc8f4c6324131f60202ef93a</sub>

tr1b0t commented on src/Tribe/Repositories/Event.php:
```diff
@@ -1635,7 +1669,8 @@ protected function order_by_venue() {
 
 		$filter_id = 'order_by_venue';
-		$this->filter_query->orderby( 'venue', $filter_id, true );
+		$order = Arr::get_in_any( [ $this->query_args, $this->default_args ], 'order', 'ASC' );
```
:warning: Equals sign not aligned with surrounding assignments; expected 5 spaces but found 1 space

<sub>hash: c51e70d193ae43929e900f086a927761</sub>

lucatume commented:
Ok, I've updated the PR to merge latest code from other PRs and fixed the tests; good for another look.

lucatume commented:
I've done more work here to fix some merge issues and a number of Month render issues (and Week in dedicated PRO PR) I've found while fixing the merge issues.

lucatume commented:
In the latest commits I've fixed the way the "this week duration" value was calculated and added more objects to the `dates` prop of the post object we return from `tribe_get_event`.

bordoni commented on src/Tribe/Models/Post_Types/Event.php:
```diff
@@ -96,56 +97,73 @@ protected function build_properties( $filter ) {
+				 * A day "crosses the EOD cutoff time" if the end is after the EOD cutoff of the start.
+				 * Here we look just for a boolean.
+				 */
+				$cross_day = tribe_end_of_day( $the_start->format( 'Y-m-d' ) ) < $the_end->format( 'Y-m-d H:i:s' );
```
Can we use the `Date_Utils` constants for these formats?

lucatume replied:
For all intents and purposes, this is a "service" format.
A few lines above that I build the `_ymd` variables using the `Ymd` format since we do not have a `Dates` constant for that, we leave that alone. But the concept is the same.

bordoni commented on src/Tribe/Views/V2/Utils/Stack.php:
```diff
@@ -349,7 +349,16 @@ protected function normalize_stack() {
+		 *
+		 * @since TBD
+		 *
+		 * @param bool $keep Whether the event should be part of the stack or not.
```
Second param for the filter.
